Deprecated WC_Cart->is_coupon_emails_allowed in favour of static DiscountsUtil::is_coupon_emails_allowed

wavvves opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR refactors code and deprecates and moves `is_coupon_emails_allowed()` from `WC_Cart` into a new ` DiscountsUtil` class as it is better suited to live there. This enables 3rd party code to use WC_Discounts in Order context, where a cart might not exist (specific use case linked in the mentioned issue below), thus preventing a fatal error from occurring when trying to access the non-existent cart.

Closes #47330 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new coupon named "exclusive", and set email restriction on it
2. Install the Code Snippets extensions and add the following snippet:
```php
add_action("wp_loaded","test_coupon",20,1);
function test_coupon(){
	WC()->cart = null; // making sure there is no cart
	$order = new WC_Order();
	$order->apply_coupon('exclusive');
}
```
3. Refresh any store page and ensure no critical error is displayed
4. Repeat the steps on an unpatched installation and verify that after adding the snippet a critical error is produced.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Fixed a fatal error when programmatically using the WC_Discounts::class in a context where no cart exists.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

wavvves committed:
Deprecated WC_Cart->is_coupon_emails_allowed in favour of WC_Discounts->is_coupon_emails_allowed

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Konamiman reviewed (dismissed):
Given that:

1. We are trying to not add any new methods to classes in `includes` (everything new should go in `src`), and
2. The code in `is_coupon_emails_allowed` is acting exclusively on the input (it doesn't depend on any state)

then I think the `is_coupon_emails_allowed` method should be a `static` method in a new `CouponUtil` class (or perhaps `DiscountUtil`) in `src/Utilities` (normally I would suggest `src/Internal/Utilities` instead, but the method is already public anyway).

Also this is a case where I would add unit tests. I can help with that if you want.

wavvves commented:
> Given that:
> 
> 1. We are trying to not add any new methods to classes in `includes` (everything new should go in `src`), and
> 2. The code in `is_coupon_emails_allowed` is acting exclusively on the input (it doesn't depend on any state)
> 
> then I think the `is_coupon_emails_allowed` method should be a `static` method in a new `CouponUtil` class (or perhaps `DiscountUtil`) in `src/Utilities` (normally I would suggest `src/Internal/Utilities` instead, but the method is already public anyway).
> 
> Also this is a case where I would add unit tests. I can help with that if you want.

I 100% agree with going the static route. I can make the change, but I do have a question. Do you think it is worth considering if, at this point, anything else should qualify for moving into a DiscountsUtil class and move it, or should I just highlight it for a future refactor?
I'm not a big fan of segmented Utility classes, as that doesn't ever sound too "clean cody" to me (maybe the object scope should be reconsidered if anything kinda falls into its responsibilities but it doesn't fit like a glove). So it becomes harder to justify opening a new Util class with just one method, that might never fit anything else in there becoming a "one trick dog" 🙄

barryhughes commented:
Adding a couple of thoughts:

> Should be a static method in a new `CouponUtil` class 
> ...
> [Should] anything else should qualify for moving into a `DiscountsUtil` class

Small thing, but if a new util class is the way we go here then, in relation to the proposed classname I also prefer `Discounts` over `Coupons` in this particular case.

> Or should I just highlight it for a future refactor?

I think this is the way to do it. I understand and agree with the concerns around segmentation, but even superficially simple refactorings tend to increase risk and so minimizing the surface area of a change has its own value (in terms of release stability).

wavvves committed:
Merge branch 'trunk' into 47330-removing-cart-dependency-from-coupon-validation

wavvves committed:
Created new DiscountsUtil class and moved is_coupon_emails_allowed()

wavvves committed:
Refactored tests.

wavvves committed:
Updated internal WC_Cart::is_coupon_emails_allowed calls and added a doing it wrong warning for 3rd party tracking.

wavvves commented:
@Konamiman @barryhughes I pushed some changes:

* Created a DiscountsUtil class w/ the static method as suggested https://github.com/woocommerce/woocommerce/pull/47589/commits/e26ca24b332630384e927b3eada1305bbb729eb5
* Moved tests that already existed https://github.com/woocommerce/woocommerce/pull/47589/commits/16d3bd114826c270ad6d025917b0a69fbace7fe3
* Took a step further, updating all calls in our codebase for the deprecated method and added a doing it wrong warning to the wrapper. I can revert this if needed https://github.com/woocommerce/woocommerce/pull/47589/commits/37d95a925e1a0f19f1bffd8e4bfcf14cfa38e58b

wavvves committed:
Linting.

barryhughes commented on plugins/woocommerce/tests/legacy/unit-tests/discounts/discounts-util.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * WC_Tests_DiscountsUtil.
+ */
+class WC_Tests_DiscountsUtil extends WC_Unit_Test_Case {
```
Thank you for adding, but can we relocate (and rename) so this lives at:

```
tests/php/src/Utilities/DiscountsUtilTest.php
```

[(See guidance here.)](https://github.com/woocommerce/woocommerce/blob/8.9.1/plugins/woocommerce/tests/README.md#guide-for-writing-unit-tests)

wavvves replied:
Thank you for pointing this out for me, I made the needed adjustments 👍🏼 
I also added a few quick asserts to the test at hand

wavvves committed:
Refactored and added new asserts to DiscountsUtilTest.

wavvves committed:
Linting.

wavvves committed:
Linting and updated test.

wavvves commented on plugins/woocommerce/tests/php/src/Utilities/DiscountsUtilTest.php:
```diff
@@ -0,0 +1,33 @@
+		$this->assertEquals( true, DiscountsUtil::is_coupon_emails_allowed( array( 'customer@wc.local' ), array( '*.local' ) ) );
+		$this->assertEquals( false, DiscountsUtil::is_coupon_emails_allowed( array( 'customer@wc.local' ), array( '*.test' ) ) );
+
+		// Case sensitivity.
```
There are [examples](https://github.com/woocommerce/woocommerce/blob/e26ca24b332630384e927b3eada1305bbb729eb5/plugins/woocommerce/includes/class-wc-discounts.php#L984) on handling this outside the method, so added this one to show that, for now, we are specifically taking case in consideration, and not something missed.

wavvves commented on plugins/woocommerce/tests/php/src/Utilities/DiscountsUtilTest.php:
```diff
@@ -0,0 +1,33 @@
+		$this->assertEquals( false, DiscountsUtil::is_coupon_emails_allowed( array(), array( 'customer@wc.local' ) ) );
+
+		// Special characters in email.
+		$this->assertEquals( false, DiscountsUtil::is_coupon_emails_allowed( array( 'special@example.com' ), array( 'special+char@example.com' ) ) );
```
Also added this one to show we're not currently considering this a match.

barryhughes commented on plugins/woocommerce/includes/class-wc-cart.php:
```diff
@@ -1696,29 +1697,18 @@ public function check_customer_coupons( $posted ) {
 	public function is_coupon_emails_allowed( $check_emails, $restrictions ) {
+		wc_doing_it_wrong(
+			'WC_Cart::is_coupon_emails_allowed',
+			'This method has been deprecated and will be removed soon. Use Automattic\WooCommerce\Utilities\DiscountsUtil::is_coupon_emails_allowed instead.',
```
We don't seem to be consistent on this across the codebase, but I think it may be best to localize this message (if a merchant notices the warning, for instance, we probably want them to see it and read it in their own language).

```suggestion
			__( 'This method has been deprecated and will be removed soon. Use Automattic\WooCommerce\Utilities\DiscountsUtil::is_coupon_emails_allowed instead.', 'woocommerce' ),
```

barryhughes reviewed (approved):
Looks good!

🗣️ Left one note about a possible i18n improvement.

🛒 Tested in the cart and checkout, using coupons with 1 allowed email, 2 allowed emails and no allowed emails, and different combinations of user data.

💻 Tested in the WP shell, with `WC()->cart` nullified, as a compliment to the provided testing instructions.

🛠️ Additional tests look great!

wavvves committed:
Added translation to wc_doing_it_wrong() as sugested.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

barryhughes commented:
Thanks! Good to merge on my end, unless we want either of the other reviewers to approve first?

wavvves commented:
> Thanks! Good to merge on my end, unless we want either of the other reviewers to approve first?

I'd be okay merging it now, but merging is blocked pending approval by @Konamiman due to requested changes
